File: problems/30/30.substring-with-concatenation-of-all-words.go
/*
 * @lc app=leetcode id=30 lang=golang
 *
 * [30] Substring with Concatenation of All Words
 */

package solution

// @lc code=start
func findSubstring(s string, words []string) []int {
	// sliding window
	// l -> keep track of start of curr substring
	// r -> iterate in increments of word_length
	// if s[r: r+word_length] is in words
	//    and cnts[word] < 0:
	// 		keep incrementing l till cnt[words] >= 0
	// for each iteration of r, check if substring equal to required substring length

	n, k, word_length := len(s), len(words), len(words[0])
	substring_size := k * word_length
	res := make([]int, 0)

	for l := 0; l < word_length; l++ {
		curr_cnts := make(map[string]int)
		for _, word := range words {
			curr_cnts[word]++
		}

		curr_l := l
		for r := curr_l; r < n-word_length+1; r += word_length {
			curr_word := s[r : r+word_length]
			curr_cnts[curr_word]--

			for curr_cnts[curr_word] < 0 {
				word_to_rem := s[curr_l : curr_l+word_length]
				curr_cnts[word_to_rem]++
				curr_l += word_length
			}

			if curr_l+substring_size == r+word_length {
				res = append(res, curr_l)
			}
		}
	}

	return res
}

// @lc code=end

File: problems/30/30.substring-with-concatenation-of-all-words.py
#
# @lc app=leetcode id=30 lang=python3
#
# [30] Substring with Concatenation of All Words
#


# @lc code=start
from collections import Counter


class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        # a concatenated substring of s is a substring that is a permutation of all words
        # return starting indices of all concatenated substrings in s
        # cannot use multiple of same word in words
        # all words are of same length
        # ie:
        # s = "barfoothefoobarman", words = ["foo","bar"]
        # i = [0,5] -> "barfoo"
        # i = [9,14] -> "foobar"
        # -> [0,9]
        # s = "wordgoodgoodgoodbestword", words = ["word","good","best","word"]
        #
        # approach:
        # l -> keep tracks of current left index of substring
        # r -> iterate range(l, n - word_length + 1, word_length)
        # if cnts[s[r: r + word_length]] < 0, keep incrementing l till cnts[s[r: r + word_length]] >= 0
        # finally check if curr substring is equal to required substring size

        n, k, word_length = len(s), len(words), len(words[0])
        substring_size = k * word_length
        res = []

        for l in range(word_length):
            cnts = Counter(words)
            print(cnts, l)
            for r in range(l, n - word_length + 1, word_length):
                curr_word = s[r : r + word_length]
                cnts[curr_word] -= 1

                while cnts[curr_word] < 0:
                    left_word = s[l : l + word_length]
                    cnts[left_word] += 1
                    l += word_length

                if l + substring_size == r + word_length:
                    res.append(l)

        return res


# @lc code=end
